### DISCUSSION TITLE
How can we use captureStream() api for fabric Canvas?

### DISCUSSION THREAD

**Discussion by HongYangWitness** [fabricjs/fabric.js#7976] [Q&A] [answered]
I saw #7187 #5873.  They used captureStream() for MediaRecorder to record the animations. I am about to implement the similar function and I don't find the api  supported in doc. And got the TypeError: canvas.CaptureStream is not a function. I am wondering how can we use it? I think since fabricjs Canvas is a kind of wrapper of original Canvas, maybe there is a way to achieve it?

![image](https://user-images.githubusercontent.com/47316888/170606473-dc018643-964b-4d5e-8d1e-af9527c13445.png)

**Comment by ShaMan123** [comment, marked as answer]
As written in https://github.com/fabricjs/fabric.js/issues/5873#issuecomment-530461722 you need to capture both canvases to capture all visuals.
You can instead use a hook `after:render` to render both canvases onto a third canvas an capture that.

**Reply by HongYangWitness** [reply]
Sorry for the ambiguity. I am not going to draw on canvas. I just want to record the animation on the fabric.js canvas, which will only have one canvas. Does that mean I still have to render the fabric.js canvas to a second original canvas (which support capturestream api) and then capture that?

**Reply by ShaMan123** [reply]
No.
`fabricCanvas.getElement().captureStream()`

**Reply by HongYangWitness** [reply]
It still gives TypeError: canvas.getElement(...).captureStream is not a function error. BTW, I am using node-canvas, because I want to do all the process (animation generation and video record) totally in the backend and without using a html.

**Reply by ShaMan123** [reply]
It's an **IMPORTANT** BTW
https://github.com/Automattic/node-canvas/issues/1698
You can implement yourself with the `after:render` hook

**Reply by HongYangWitness** [reply]
Thank you!

**Reply by HongYangWitness** [reply]
**Update: After I switch it to fabricCanvas and using fabricCanvas.getElement().captureStream(), it still gives me  TypeError: canvas.getElement(...).captureStream error.
<img width="398" alt="image" src="https://user-images.githubusercontent.com/47316888/170775583-d90f8427-f93b-432f-ac40-3c9a8caa72ad.png">
<img width="584" alt="image" src="https://user-images.githubusercontent.com/47316888/170775675-0465e8f4-2903-4680-ba6f-233883c94dd8.png">
